### ISSUE TITLE
Template arguments are incorrectly treated as forward declarable

### ISSUE THREAD

**Issue by shivanshu3** [include-what-you-use/include-what-you-use#796]
foo.h:
```C++
struct Foo
{};
```

main.cpp:
```C++
#include "foo.h"

template<typename T>
struct TypeContainer
{
    typedef T type;
};

int dummy()
{
    return sizeof(TypeContainer<Foo>::type);
}
```

IWYU output:
```text
main.cpp should add these lines:
struct Foo;

main.cpp should remove these lines:
- #include "foo.h"  // lines 1-1

The full include-list for main.cpp:
struct Foo;
---
```

After doing some debugging, looks like making the function `DetermineForwardDeclareStatusForTemplateArg` a no-op fixes the bad behavior.

I can repro this behavior on the branch `clang_10` at commit `a40a28740c963636d8d41e75f297e8e078302d70`

**Comment by shivanshu3** [comment]
Note that the following tests start to fail after I make `DetermineForwardDeclareStatusForTemplateArg` a no-op:

```text
alias_template
casts
default_template_arg_other_file
derived_function_tpl_args
explicit_instantiation2
funcptrs
precomputed_tpl_args
sizeof_reference
template_args
typedef_in_template
virtual_tpl_method
```

**Comment by kimgr** [comment]
Could you take a look at this branch: https://github.com/include-what-you-use/include-what-you-use/pull/775/?

I think @jbytheway has gone to great lengths to fix up template argument handling there. It would be interesting to see if it somehow fixes this even if it doesn't touch `DetermineForwardDeclareStatusForTemplateArg` directly. The test sizeof case looks eerily similar to your repro above.

**Comment by shivanshu3** [comment]
I tried applying the patch from that PR and my test still fails.
@kimgr 
@jbytheway FYI

**Comment by jbytheway** [comment]
Yeah, I discovered when I came back to #775 this weekend that this case is still not handled correctly.  I don't plan to fix it on that branch.  Not sure how it would best be fixed at all...

**Comment by jbytheway** [comment]
Now that #775 is merged, I just wanted to come back to this issue and write down my thoughts on it while I had them.

The way IWYU currently determines whether a template argument use is a full-use is by instantiating that template.  Each use of the arg within the template is checked for full-use-ness.  This process often involves recursively instantiating many templates, but IWYU tracks the original template arguments to the top-level template to find any full-uses thereof.

However, once that recursive instantiation process is done, IWYU forgets everything about the template.

In an example like yours `TypeContainer<Foo>` is instantiated, and `Foo` is only used as a typedef, so the argument is not full-used.

Then `TypeContainer<Foo>::type` is full-used, but at this point there's no way to link that back to the template argument of `TypeContainer<Foo>`.  You may say it's obvious that they're linked, because they're the same type, but that could just be a coincidence.  Perhaps `Foo` is an implementation detail of `TypeContainer` as well as being a template argument.

I think the only way to solve this problem in general is to remember some more information about template instantiations.  In this case we'd need to remember that `TypeContainer<Foo>::type` is specifically defined using the argument to `TypeContainer` and thus full-uses thereof should be ascribed to the code using that member typedef rather than assumed an implementation detail.

In this particular case that's not too bad, but if you consider a more complex example, like:

```c++
using map = std::map<Foo, Bar>;
return sizeof(map::value_type::first_type);
```

which should be a full-use of Foo but (maybe) not Bar, it's not entirely clear to me exactly how one would track all the relevant connections.  I think it's possible, but it's going to be complicated.

I don't anticipate attempting this myself unless I actually discover the necessity in my own use of IWYU.

**Comment by kimgr** [comment]
@jbytheway Thank you for the logical and systematic explanation!

**Comment by kimgr** [comment]
This appears to be fixed on master. Closing.
